### PR TITLE
Refactor integrations logging

### PR DESCRIPTION
**Pull request by GGP1**
|Related issue|
|---|
| Closes https://github.com/wazuh/wazuh/issues/15205 |

> IMPORTANT: this pull request should be merged together with the one from the issue https://github.com/wazuh/wazuh/issues/16662

## Description

Improve the level of detail of the information logged during integrations executions. New levels have been added so the user can specify how much and what type of information he wants to be displayed.

A new help message is shown when the arguments provided by the user are not valid. Using file paths that do not exist or an invalid URL for the webhook now throws an error.

## Logs/alerts

Examples shown were captured while executing the Slack integration.

#### Logging level: `INFO`

Before

```
```

Now

```
Wed Apr 05 11:47:38 UTC 2023 INFO Starting
Wed Apr 05 11:47:38 UTC 2023 INFO Alerts file location: alerts.json
Wed Apr 05 11:47:38 UTC 2023 INFO Webhook: https://hooks.slack.com/services/XXXXXX/XXXXX/XXXXX
Wed Apr 05 11:47:38 UTC 2023 INFO Processing alert with ID 1680616767.732107
Wed Apr 05 11:47:38 UTC 2023 INFO Generating message
Wed Apr 05 11:47:38 UTC 2023 INFO Sending message
Wed Apr 05 11:47:38 UTC 2023 INFO Message sent successfully
```

#### Logging level: `DEBUG`

Before

```
Wed Apr 05 11:55:12 UTC 2023: # Starting
Wed Apr 05 11:55:12 UTC 2023: # Webhook
Wed Apr 05 11:55:12 UTC 2023: https://hooks.slack.com/services/XXXX/XXXX/XXXX
Wed Apr 05 11:55:12 UTC 2023: # File location
Wed Apr 05 11:55:12 UTC 2023: # alerts.json
Wed Apr 05 11:55:12 UTC 2023: # Processing alert
Wed Apr 05 11:55:12 UTC 2023: {'timestamp': '2023-04-04T13:59:27.021+0000', 'rule': {'level': 3, 'description': 'Ossec server started.', 'id': '502', 'firedtimes': 1, 'mail': False, 'groups': ['ossec'], 'pci_dss': ['10.6.1'], 'gpg13': ['10.1'], 'gdpr': ['IV_35.7.d'], 'hipaa': ['164.312.b'], 'nist_800_53': ['AU.6'], 'tsc': ['CC7.2', 'CC7.3']}, 'agent': {'id': '000', 'name': 'wazuh-master'}, 'manager': {'name': 'wazuh-master'}, 'id': '1680616767.732107', 'cluster': {'name': 'wazuh', 'node': 'master-node'}, 'full_log': 'ossec: Manager started.', 'decoder': {'name': 'ossec'}, 'location': 'wazuh-monitord'}
Wed Apr 05 11:55:12 UTC 2023: # Generating message
Wed Apr 05 11:55:12 UTC 2023: {"attachments": [{"color": "good", "pretext": "WAZUH Alert", "title": "Ossec server started.", "text": "ossec: Manager started.", "fields": [{"title": "Agent", "value": "(000) - wazuh-master"}, {"title": "Location", "value": "wazuh-monitord"}, {"title": "Rule ID", "value": "502 _(Level 3)_"}], "ts": "1680616767.732107"}]}
Wed Apr 05 11:55:12 UTC 2023: # Sending message
Wed Apr 05 11:55:12 UTC 2023: {response object}
```

Now

```
Wed Apr 05 11:55:12 UTC 2023 DEBUG alerts.json - https://hooks.slack.com/services/XXXX/XXXX/XXXX DEBUG
Wed Apr 05 11:55:12 UTC 2023 INFO Starting
Wed Apr 05 11:55:12 UTC 2023 INFO Alerts file location: alerts.json
Wed Apr 05 11:55:12 UTC 2023 INFO Webhook: https://hooks.slack.com/services/XXXX/XXXX/XXXX
Wed Apr 05 11:55:12 UTC 2023 INFO Processing alert with ID 1680616767.732107
Wed Apr 05 11:55:12 UTC 2023 DEBUG Alert:
{'timestamp': '2023-04-04T13:59:27.021+0000', 'rule': {'level': 3, 'description': 'Ossec server started.', 'id': '502', 'firedtimes': 1, 'mail': False, 'groups': ['ossec'], 'pci_dss': ['10.6.1'], 'gpg13': ['10.1'], 'gdpr': ['IV_35.7.d'], 'hipaa': ['164.312.b'], 'nist_800_53': ['AU.6'], 'tsc': ['CC7.2', 'CC7.3']}, 'agent': {'id': '000', 'name': 'wazuh-master'}, 'manager': {'name': 'wazuh-master'}, 'id': '1680616767.732107', 'cluster': {'name': 'wazuh', 'node': 'master-node'}, 'full_log': 'ossec: Manager started.', 'decoder': {'name': 'ossec'}, 'location': 'wazuh-monitord'}
Wed Apr 05 11:55:12 UTC 2023 INFO Generating message
Wed Apr 05 11:55:12 UTC 2023 DEBUG Message:
{"attachments": [{"color": "good", "pretext": "WAZUH Alert", "title": "Ossec server started.", "text": "ossec: Manager started.", "fields": [{"title": "Agent", "value": "(000) - wazuh-master"}, {"title": "Location", "value": "wazuh-monitord"}, {"title": "Rule ID", "value": "502 _(Level 3)_"}], "ts": "1680616767.732107"}]}
Wed Apr 05 11:55:12 UTC 2023 INFO Sending message
Wed Apr 05 11:55:13 UTC 2023 INFO Message sent successfully
Wed Apr 05 11:55:13 UTC 2023 DEBUG Slack response:
Date: Wed, 05 Apr 2023 11:55:12 GMT
Status code: 200
URL: https://hooks.slack.com/services/XXXX/XXXX/XXXX
Slack unique ID: XXX
```

> We may prefer to use one line per message, in that case, we should encode the objects first

#### Invalid arguments

Before

> Only if debug mode was enabled, if it was not, no message was shown.

```
Wed Apr 05 11:55:12 UTC 2023: # Exiting: Bad arguments.
```

Now

```
Exiting: Invalid arguments.
    
Usage:
    slack <alerts_file> [api_key] <webhook_url> [logging_level]

Arguments:
    alerts_file (required)
        Path to the JSON file containing the alerts.

    api_key (not required)
        The API key argument is not needed for the Slack integration. However, it's still considered because the 
        integrator executes all scripts with the same arguments.
        If you are executing the script manually, please put anything in that argument.

    webhook_url (required)
        Slack webhook URL where the messages will be sent to.

    logging_level (optional)
        Used to define how much information should be logged. Default is INFO.

        Levels: NOTSET, DEBUG, INFO, WARNING, ERROR, CRITICAL.
    
```

## Tests

```console
(venv) gasti@pop-os:~/work/wazuh/integrations$ python3 -m pytest . --disable-warnings
================================================================= test session starts =================================================================
platform linux -- Python 3.9.16, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /home/gasti/work/wazuh/integrations
plugins: html-3.1.1, testinfra-5.0.0, metadata-2.0.4
collected 95 items                                                                                                                                    

tests/test_maltiverse.py .............................................                                                                          [ 47%]
tests/test_pagerduty.py ..........                                                                                                              [ 57%]
tests/test_shuffle.py ..................                                                                                                        [ 76%]
tests/test_slack.py ..........                                                                                                                  [ 87%]
tests/test_virustotal.py ............                                                                                                           [100%]

================================================================= 95 passed in 0.20s ==================================================================
```

### CODE REVIEW

File: integrations/slack
Comment (GGP1) [inline integrations/slack:120]: Perhaps we should encode the response and log it as is, adding multiline messages may make the log be harder to read. Let me know if you agree or if it's OK as it is.

**Review by RamosFe — DISMISSED** [review]

File: integrations/slack
Comment (EduLeon12) [reply integrations/slack:120]: Encoding the JSON will not necessarily help in reducing the size of the log file compression will be needed after the encoding, and since that will require a decompression for reading it may cause complexity on the current issue.

It's a good subject to be aware of.

**Review by EduLeon12 — DISMISSED** [review]
After testing integrations they all  work as described in the issue.

LGTM !!

**Comment by davidjiglesias** [comment]
Good job @GGP1,

However, there are a lot of changes in this PR, and we have very little to no testing for this part of the product. 

For that reason, and in order to approve this and merge the PR, we should do the following:

- Run virustotal E2E UX test for this branch (requires issue)
- Run shuffle + slack + pagerduty E2E UX test for this branch (requires issue)
- Approval of the PR and code review from @wazuh/framework manager
- Approval of the PR and code review from someone in @wazuh/core

**Review by EduLeon12 — DISMISSED** [review]
Approved after conclusion explained [here](https://github.com/wazuh/wazuh/issues/15205#issuecomment-1658408791).

LGTM !

File: integrations/slack
Comment (Selutario) [reply integrations/slack]: I agree with converting the multiline log into a JSON. This can make it easier to create rules for users who want to monitor this log file.

File: integrations/slack.py
Comment (Selutario) [inline integrations/slack.py:302]:
```suggestion
        return False


def setup_logger(args):
```

File: integrations/slack.py
Comment (Selutario) [inline integrations/slack.py:291]:
```suggestion
    help_msg = '''
    Exiting: Invalid arguments.
    
    Usage:
        slack <alerts_file> [api_key] <webhook_url> [logging_level] [options_file]
    Arguments:
        alerts_file (required)
            Path to the JSON file containing the alerts.
        api_key (not required)
            The API key argument is not needed for the Slack integration. However, it's still considered because the 
            integrator executes all scripts with the same arguments.
            If you are executing the script manually, please put anything in that argument.
        webhook_url (required)
            Slack webhook URL where the messages will be sent to.
        logging_level (optional)
            Used to define how much information should be logged. Default is INFO.
            Levels: NOTSET, DEBUG, INFO, WARNING, ERROR, CRITICAL.
        options_file (optional)
            Path to a file containing custom variables to be used in the integration. It must be JSON-encoded.
    '''
```

File: integrations/slack.py
Comment (Selutario) [inline integrations/slack.py:303]: Missing docstring

File: integrations/slack.py
Comment (Selutario) [inline integrations/slack.py:296]: Missing docstring

File: integrations/shuffle.py
Comment (Selutario) [inline integrations/shuffle.py:76]: Same as for slack, I would move this to the `else` clause and remove the `invalid_arguments` variable.

File: integrations/shuffle.py
Comment (Selutario) [inline integrations/shuffle.py:98]:
```suggestion
    """Create a message with all valid fields.
```

File: integrations/shuffle.py
Comment (Selutario) [inline integrations/shuffle.py:280]:
```suggestion
    """Send the command's help message to the standard output."""
```

File: integrations/shuffle.py
Comment (Selutario) [inline integrations/shuffle.py:299]:
```suggestion
    help_msg = '''
    Exiting: Invalid arguments.
    
    Usage:
        shuffle <alerts_file> [api_key] <webhook_url> [logging_level] [options_file]
    Arguments:
        alerts_file (required)
            Path to the JSON file containing the alerts.
        api_key (not required)
            The API key argument is not needed for the Shuffle integration. However, it's still considered because the 
            integrator executes all scripts with the same arguments.
            If you are executing the script manually, please put anything in that argument.
        webhook_url (required)
            Shuffle webhook URL where the messages will be sent to.
        logging_level (optional)
            Used to define how much information should be logged. Default is INFO.
            Levels: NOTSET, DEBUG, INFO, WARNING, ERROR, CRITICAL.
        options_file (optional)
            Path to a file containing custom variables to be used in the integration. It must be JSON-encoded.
    '''
```

File: integrations/shuffle.py
Comment (Selutario) [inline integrations/shuffle.py:304]:
```suggestion
    """Validate a URL.
```

File: integrations/shuffle.py
Comment (Selutario) [inline integrations/shuffle.py:323]:
```suggestion
    """Configure the logger.
```

File: integrations/virustotal.py
Comment (Selutario) [inline integrations/virustotal.py:73]: Move to else clause.

File: integrations/virustotal.py
Comment (Selutario) [inline integrations/virustotal.py:93]:
```suggestion
    """This is the core function, creates a message with all valid fields
    and overwrite or add with the optional fields
```
```suggestion
    """Create a message with all valid fields and overwrite or add the optional fields.
```

File: integrations/virustotal.py
Comment (Selutario) [inline integrations/virustotal.py:98]:
```suggestion
        The argument list from main call.
```

File: integrations/virustotal.py
Comment (Selutario) [inline integrations/virustotal.py:117]:
```suggestion
    """Generate the JSON object with the message to be sent.
```

File: integrations/virustotal.py
Comment (Selutario) [inline integrations/virustotal.py:127]:
```suggestion
        The JSON message to send.
```

File: integrations/virustotal.py
Comment (Selutario) [inline integrations/virustotal.py:183]:
```suggestion
    """Send a request to VT API and fetch information to build message.
```

File: integrations/virustotal.py
Comment (Selutario) [inline integrations/virustotal.py:195]:
```suggestion
    hash : str
        Hash need it for parameters.
    apikey: str
        Authentication API.

    Returns
    -------
    data: any
        JSON with the response.
```

File: integrations/virustotal.py
Comment (Selutario) [inline integrations/virustotal.py:255]:
```suggestion
    """Read JSON alert object from file.
```

File: integrations/virustotal.py
Comment (Selutario) [inline integrations/virustotal.py:284]: You included a comment for this function in other integrations, I would also include it here.

File: integrations/virustotal.py
Comment (Selutario) [inline integrations/virustotal.py:303]:
```suggestion
    Exiting: Invalid arguments.
    
    Usage:
        virustotal <alerts_file> [api_key] <webhook_url> [logging_level] [options_file]
    Arguments:
        alerts_file (required)
            Path to the JSON file containing the alerts.
        api_key (required)
            Virus total API key.
        webhook_url (not required)
            The webhook URL argument is not needed for the Virus total integration. However, it's still considered because the 
            integrator executes all scripts with the same arguments.
            If you are executing the script manually, please put anything in that argument.
        logging_level (optional)
            Used to define how much information should be logged. Default is INFO.
            Levels: NOTSET, DEBUG, INFO, WARNING, ERROR, CRITICAL.
        options_file (optional)
            Path to a file containing custom variables to be used in the integration. It must be JSON-encoded.
    '''
```

File: integrations/virustotal.py
Comment (Selutario) [inline integrations/virustotal.py:307]: Missing docstring

**Review by Selutario — CHANGES_REQUESTED** [review]
Most of the requested changes are small typos. However, there are some bigger things to take into consideration:

## Log rotation
Now all integrations will print logs for each alert and by default. This can cause the `integration.log` file to grow large. It is necessary to study if it rotates automatically based on some criteria. If it doesn't, it would have to be implemented in some way and verify that there are no problems in the integrations during the rotation.

## Maltiverse integration
A new integration was added in 4.7.0 and will probably require refactoring its logging tool too.

## Logging secrets
I'm not entirely sure if its a good practice to print things like API keys/tokens and other secrets in a log file. Maybe we should hide that info by not logging it or replacing it with `****`, although it can be counterproductive for troubleshooting purposes. If possible, do some research on it.

## E2E Testing
Once all the mentioned above is solved, we should finally do some E2E testing in separated child issues as requested here before merging:
- https://github.com/wazuh/wazuh/pull/16624#issuecomment-1568207142

File: integrations/pagerduty.py
Comment (GGP1) [reply integrations/pagerduty.py:117]: The message is built by the script if the arguments are correct, if they are not, then we fail. So an empty message will never reach this point.

File: integrations/maltiverse.py
Comment (GGP1) [inline integrations/maltiverse.py:646]: Set a `backupCount` of 31 based on the default value used for the other logs: https://github.com/wazuh/wazuh/blob/2960662dcf23c88ea73eefdd2c330028426957aa/etc/internal_options.conf#L280-L281

**Pull request closed without merging**
